packages/sidenav/src/index.tsx

```diff
@@ -0,0 +1,381 @@
+/** @jsx jsx */
+import { jsx, ThemeProvider, SxStyleProp, Theme } from 'theme-ui'
+import { MDXProvider, MDXProviderComponents } from '@mdx-js/react'
+import React, {
+  useState,
+  FunctionComponent,
+  EventHandler,
+  MouseEvent,
+  forwardRef,
+  ReactNode,
+  ReactComponentElement,
+  ReactElement,
+} from 'react'
+import { Global } from '@emotion/core'
+import merge from 'deepmerge'
+
+const createNestedLinks = (
+  children: ReactNode,
+  depth: number = 0
+): ReactNode[] => {
+  const links = React.Children.toArray(children).reduce(
+    (acc: ReactNode[], child) => {
+      if (!React.isValidElement(child)) {
+        return acc
+      }
+      const type: string = child.props && child.props.mdxType
+      if (!child.props || !child.props.children) return acc
+      if (type === 'a') return [...acc, child]
+      if (depth > 0 && type === 'ul') {
+        const last = acc[acc.length - 1]
+        acc[acc.length - 1] = React.cloneElement(last as ReactElement, {
+          links: createNestedLinks(child.props.children),
+        })
+        return acc
+      }
+      return [...acc, ...createNestedLinks(child.props.children, depth + 1)]
+    },
+    []
+  ) as ReactNode[]
+  return links
+}
+
+const flattenLinks = (children: ReactNode): ReactNode[] =>
+  React.Children.toArray(children).reduce((acc: ReactNode[], child) => {
+    if (React.isValidElement(child)) {
+      if (child.props && child.props.mdxType === 'a') {
+        return [...acc, child]
+      }
+      if (!child.props || !child.props.children) return acc
+      return React.Children.toArray([
+        ...acc,
+        ...flattenLinks(child.props.children),
+      ])
+    }
+    return acc
+  }, []) as ReactNode[]
+
+const Overlay: FunctionComponent<{
+  onClick?: EventHandler<MouseEvent<HTMLDivElement>>
+}> = ({ onClick }) => (
+  <React.Fragment>
+    <div
+      onClick={onClick}
+      sx={{
+        position: 'fixed',
+        top: 0,
+        right: 0,
+        bottom: 0,
+        left: 0,
+      }}
+    />
+    <Global
+      styles={{
+        body: {
+          overflow: ['hidden', 'auto'],
+        },
+      }}
+    />
+  </React.Fragment>
+)
+
+export const Sidenav = forwardRef<
+  HTMLDivElement,
+  {
+    open?: boolean
+    components: MDXProviderComponents
```

Comment:
This was previously treated as `MDXProviderComponents | undefined`, right?
Could we default to empty object or even leave undefined if it's okay with MDXProvider types?

Reply:
Yeah and that was probably an oversight. Making components optional / allowing undefined will cause a type-check to fail with MDXProvider that requires components to be defined.

<img width="1142" alt="Screenshot 2020-05-16 at 09 10 13" src="https://user-images.githubusercontent.com/1410207/82114523-15160380-9755-11ea-82ff-01b9d9d0b574.png">

Reply:
We can default to `{}` to avoid a breaking change.

Reply:
Updated